packages/alfa-rules/src/sia-r19/rule.ts

```diff
@@ -97,3 +105,59 @@ function isValid(attribute: aria.Attribute): boolean {
       return value.split(/\s+/).every((value) => options.some(equals(value)));
   }
 }
+
+/**
+ * All ids that exist in the tree of a given root.
+ */
+const idsCache = Cache.empty<Node, Set<string>>();
+
+function treeHasId(id: string, node: Node): boolean {
+  // We absolutely need no traversal options here, because `id` are scoped to
+  // trees, so we do not want to cross shadow or content boundaries.
+  return idsCache
+    .get(node.root(), () =>
+      // Turning it into a Set for faster lookup.
+      Set.from(
+        node
+          .root()
+          .descendants()
+          .filter(isElement)
+          .collect((element) => element.id)
+      )
+    )
+    .has(id);
+}
+
+function isRequiredRefValid(
```

Comment:
nitpick: perhaps would make sense to be less specific since we're handling also other-than required reference attributes?
```suggestion
function isAttributeValid(
```

Reply:
Struggled a bit to find a correct name 😅 At list the doc should make the intention explicit…